### DISCUSSION TITLE
Pagination & Actions for selections

### DISCUSSION THREAD

**Discussion by mayankjanmejay** [bosskmk/pluto_grid#148] [Ideas]
Hi,

I love this plugin but I have two suggestions for improvements.

1) Pagination -> It would be awesome to have pagination built into the table.
2) Actions -> It would also be awesome if we could input our own widgets for actions like Delete or others...

Thank you :)

**Comment by jeansebastienZ** [comment]
Hello , 

You can add any action you want in grid . You simply need to use Cell renderer : https://weblaze.dev/pluto_grid/build/web/#feature/cell-renderer

And add a widget like icon or image , then detect the onclick with an Inkwell and here you go !
If you need to detect the whole cell click or whatever you can use statemanager :) .

**Reply by mayankjanmejay** [reply]
Awesome, thank you for the very very much for the advice, I will use that now 😃 .

How about pagination and somewhere to display the number of selected rows? Do you think we could get these features in an upcoming update?

**Comment by jeansebastienZ** [comment]
I think you can handle the pagination manually with some variable but ya need to dig in the way you handle your rowsvalue . 
It shouldnt be complicated . 
For any upcoming update however , i the owner of the package should let you know !

**Reply by mayankjanmejay** [reply]
That is very true, while its not a difficult thing to add in but it is definitely a nice feature to have out of the box 🥡 .

I do not understand your comment where you mention "ya need to dig in the way you handle your rowsvalue ." Could you please elaborate?

Sure yea, It would be awesome of you to announce the features you are working on. Do you have a page or list somewhere of the upcoming features? 😃

**Comment by jeansebastienZ** [comment]
@bosskmk  is the owner of the package , sure it would be interesting to see whats coming soon .  Ok , for the pagination , let me few days i let you some example , could you just show here visually how you would it need to be ?

**Reply by mayankjanmejay** [reply]
How ever you like, I made the pagination this way. I put the page selection in the footer and I move the start index according to the page. I will also be adding a dropdown to selection the number of items per page again in the footer.

Opps, sorry from your wording I understood you were the package creator. My mistake 🙅.


![image](https://user-images.githubusercontent.com/35850067/107544437-e4bfa080-6bd2-11eb-9c78-5282b364e2f7.png)

**Reply by jeansebastienZ** [reply]
Basically for now thats the best option till @bosskmk  implement something like this . But i'm not even sure it's usefull as you can basically do it by yourself for now . But still a future update with this could be interesting.

**Reply by mayankjanmejay** [reply]
I am having a very weird problem. 
This is my code for the pagination it works to put rows into the var but then the var does not update in the UI.
 
Even if the var gets updated to no records the UI still shows the same records as it did on load.

![image](https://user-images.githubusercontent.com/35850067/107548902-dde75c80-6bd7-11eb-8775-6b00093619f1.png)

![image](https://user-images.githubusercontent.com/35850067/107548933-ea6bb500-6bd7-11eb-815d-fcfc69835eb9.png)

Is there something I am doing wrong?


```
class MyProjects extends StatefulWidget {

  final List<JProject> projects;
  static const routeName = 'feature/value-formatter';

  const MyProjects({Key key, this.projects}) : super(key: key);

  @override
  _MyProjectsState createState() => _MyProjectsState();
}

class _MyProjectsState extends State<MyProjects> {

  int startIndex=0;
  int itemsPerPage=10;
  int pageNo=0;
  List<PlutoColumn> columns;

  List<PlutoRow> rows = new List();

  @override
  void initState() {
    super.initState();
    log.shout('MyProjects Loading Data');
    columns = [
      PlutoColumn(
        title: 'Project ID',
        field: 'projectNo',
        enableRowChecked: false,
        enableEditingMode: false,
        sort: PlutoColumnSort.descending,
        type: PlutoColumnType.number(),    
       
      ),
      PlutoColumn(
        title: 'Project Name',
        field: 'projectName',
        type: PlutoColumnType.text(          //readOnly: true,
        ),
        applyFormatterInEditing: false,
        enableEditingMode: false,
        enableContextMenu: false
       ),
      PlutoColumn(
        title: 'Project Date',
        field: 'projectDate',
        type: PlutoColumnType.date(),
        enableEditingMode: false,
        applyFormatterInEditing: false,
      ),
      PlutoColumn(
        title: 'Actions',
        field: 'actions',
        enableRowChecked: false,
        enableEditingMode: false,
        //sort: PlutoColumnSort.descending,
        type: PlutoColumnType.text(),    
        renderer: (rendererContext) {
         return new Row(children: [
            new IconButton(icon: Icon(Icons.delete), onPressed: () {
              
            },)
          ],);
        },
      ),

    ];
   

  }

  @override
  Widget build(BuildContext context) {
 if (widget.projects!=null)
    {
rows.clear();
      for (startIndex= startIndex; startIndex < (widget.projects.length> pageNo*itemsPerPage ?widget.projects.length :pageNo*itemsPerPage); startIndex++) {
        rows.add(PlutoRow(
                            cells: {
                              'projectNo': PlutoCell(value: widget.projects[startIndex].projectNo),
                              'projectName': PlutoCell(value: widget.projects[startIndex].projectName),
                              'projectDate': PlutoCell(value: widget.projects[startIndex].projectDate.toIso8601String()),
                              'actions': PlutoCell(value: widget.projects[startIndex].projectDate.toIso8601String()),
                            },
      ));
      }
    }
    return  PlutoGrid(
      mode: PlutoGridMode.select,
        columns: columns,
        rows: rows,
        configuration:PlutoGridConfiguration  (
          localeText: PlutoGridLocaleText(),
          //columnFilterConfig: PlutoGridColumnFilterConfig(),
          scrollbarConfig: PlutoGridScrollbarConfig()
        ),
        createFooter: (PlutoGridStateManager grid) {

          List<Widget> rowzz= new List();
          num maxCount=widget.projects.length/itemsPerPage;

          for (num i = 0; i < (maxCount); i++) {

            rowzz.add(new FlatButton(onPressed: () { 
              setState(() {
                if (i==0)
                {
                  pageNo=0;
                  startIndex=0;
                }
                else 
                {
                  pageNo=i;
                  startIndex= (i)*itemsPerPage;
                }
                

                log.shout(startIndex);
              });
             },
            child: new Text('$i'), ));
            
          }
          return new Row(children: rowzz,);
        },
        onLoaded: (event) {
          
        },
        onSelected: (PlutoGridOnSelectedEvent event) {
          log.shout(event.cell);
          print(event);
        },
        onChanged: (PlutoGridOnChangedEvent event) {
          log.shout(event);
          print(event);
        },
    );
  }
}
```

**Comment by jeansebastienZ** [comment]
Hello , don't get lot of times to check out all of your codes but what i can advice you if you didnt ( i got same kind of problem last time ) .

First -> i stopped used clear with pluto rows , seems it keep the row in memory instead of 
  `  plutorows.clearData();`
just use 
`    plutorows = List();`

Second - > to clear the rows from plutogrid  stateManager.removeRows(plutorows);
Third - >  stateManager.appendRows(plutorows); // add your ros
Fourth -> one example i use to refresh rows  :  
```
 
    plutorows = List();
    stateManager.removeRows(plutorows);
 
      final bool dataLoaded =  await _ordreReparationCon.refreshData(); // load all date from Backoffice

    if (dataLoaded) {
      loadAllPlutoColumn(); // load your column
      loadAllPlutorows();// load your rows
    }
    setState(() {
      stateManager.appendRows(plutorows);
    });
```
Hope it help you .

**Reply by mayankjanmejay** [reply]
Hi, 

I don't have statemanager a variable. How do I instantiate this var?

I also tried the trick you mentioned of putting Widgets in the Column renderer using that I am not able to embed something like list.removeAt(index) or other row specific commands for which I need the index or some cell data to do the operation on.
